### ISSUE TITLE
Add support for named loggers

### ISSUE THREAD

**Issue by creiht** [uber-go/zap#130]
It can often be convenient for logs to have a name that is part of each log line.  For example, if multiple systems are writing to the same log file, or for pastes of log messages to make it obvious what system the log line came from.

I would like to propose adding a name argument to the logger.New and logger.With methods.  When a child log is created the two log names are concatenated together with a '.'. 

So for the example in the docs:

`{"level":"warn","msg":"Log without structured data..."}`

Would be something like:

`{"name":"myapp","level":"warn","msg":"Log without structured data..."}`

or for a child logger:

`{"name":"myapp.child","level":"warn","msg":"Log without structured data..."}`

**Comment by magicalbanana** [comment]
Why not just expand the usage of the context?

**Comment by creiht** [comment]
@magicalbanana: That's what we are currently doing, and it is workable, but has a couple of problems.  The biggest being that the behavior with multiple keys for a single name is undefined.  So for example, currently if you create a log with a name field of "myapp", and then create a child log with a name field of "myapp.child", you will get a log entry with both name fields.  It would be more convenient to be a fundamental part of the log information.

**Comment by akshayjshah** [comment]
Hey @creiht - interesting idea. What you're proposing is certainly doable, but I'm curious what you think about two alternatives:

1. What do you think of #132? Separate components or systems would log to separate portions of the top-level namespace.
2. Could this be handled automatically by the `AddCaller` hook? I.e., is including a file name and line number sufficient to identify separate components, or does the user have to identify the module hierarchy?

**Comment by creiht** [comment]
@akshayjshah:

1. I think that would certainly help, and I'm cool with trying that first.
2. The problem is that separate systems could share code, so in our case we still need a way to differentiate which system is sending the log.

**Comment by akshayjshah** [comment]
Ah, good point about the second option - I didn't consider logging from shared utility code.

**Comment by heavenlyhash** [comment]
Fwiw, I just came to a similar situation, and also badly want to be able to replace a named context field as I fork a new child logger.

I looked at #132 as well, and that's definitely an awesome feature, but really doesn't do what I need -- I have another part of the system that has already defined hierarchical names for modules of the system, and I just want to plug and chug those unmolested.  So replacing a named field is exactly the right thing to do.

**Comment by prashantv** [comment]
I think hierarchical names would be useful to quickly identify the component that caused a log. Unfortunately replacing arbitrary fields isn't going to be possible. I think names would have to be a top-level construct in zap, with new APIs:
- `logger.Named(string)` to create a logger that adds a new name
- `Entry` would need a new `Name` field.

I'd expect the result to be a single name joined by a something like `.` (possibly customizable).

**Comment by akshayjshah** [comment]
Unless we can come up with a good, encoding-agnostic approach to #132, adding names as a top-level concern makes sense to me.

**Comment by rogpeppe** [comment]
I was surprised at the fact that With doesn't replace existing fields with the
same name - that rules out a significant space of possibilities where we'd
want a value that reflects the current context but is overridden by children
(one example I have is a UUID pertaining to an internal RPC request).

> Unfortunately replacing arbitrary fields isn't going to be possible.

I wonder if it's as impossible as you say. What if With kept the fields
around in the logger as well as encoding them immediately?
Then there could be a new operation in Logger, say Replace:

    // Replace is like With except that any values with keys that
    // are the same as existing keys will be replaced rather than added.
    Replace(fields ...Field) Logger

Replace would go through all of the current fields associated with the
logger and change them if it finds any the same, then add them all
to an Encoder instance freshly cloned from the original encoder
that the logger was created with.

I think this would mean that With would incur one more allocation (the
fields slice would need to be copied). The semantics of Replace would
also be slightly different from With in that the previous With values would
be encoded again, potentially giving different results, although it's
probably bad practice to be relying on that.

**Comment by akshayjshah** [comment]
I'm reluctant to add more overhead to `With` - the overhead of copying maps around was one of the main reasons we moved away from Logrus. However, I think that the `Replace` method is a possibility with some encoder support; rather than copying the fields from parent to child logger, we could have the encoders implement a `Delete` method that snips out a portion of the accumulated bytes. The logger would then implement `Replace` as a delete & add. Opened #192 to track this.

In the short term, though, most JSON decoders I tested with (including Go's `encoding/json`) keep the last value for a given key.
